src/core/matchers/matchersUtil.js

```diff
@@ -239,21 +239,66 @@ getJasmineRequireObj().matchersUtil = function(j$) {
       if (!result) {
         return false;
       }
-    } else if (className == '[object Set]' || className == '[object Map]') {
+    } else if (className == '[object Map]') {
       if (a.size != b.size) {
         diffBuilder.record(a, b);
         return false;
       }
-      var iterA = a.entries(), iterB = b.entries();
-      var valA, valB;
-      do {
-        valA = iterA.next();
-        valB = iterB.next();
-        if (!eq(valA.value, valB.value, aStack, bStack, customTesters, j$.NullDiffBuilder())) {
-          diffBuilder.record(a, b);
-          return false;
+
+      // For both sets of keys, check they map to equal values in both maps
+      var mapKeys = [a.keys(), b.keys()];
+      var mapIter, mapKeyIt, mapKey;
+      for (i = 0; result && i < mapKeys.length; i++) {
+        mapIter = mapKeys[i];
+        mapKeyIt = mapIter.next();
+        while (result && !mapKeyIt.done) {
+          mapKey = mapKeyIt.value;
+          result = eq(a.get(mapKey), b.get(mapKey), aStack, bStack, customTesters, j$.NullDiffBuilder());
```

Comment:
I did actually give `diffBuilder.withPath()` while writing this and the output looked odd to me (the `$` was throwing me off), but reading more of the tests now I can see that's the expected error syntax.

However IIUC the ObjectPath used by DiffBuilder only supports strings or numbers as path components, but Map keys can actually be anything, so even assuming we extend ObjectPath, I'm not sure how readable the path would be if a key is a complex object?

I guess there are different options:
1. Just report that the Map/Set are different, as per the current code (noop).
2. Use `diffBuilder.withPath()` iff keys are numbers or strings, and fallback to 1. otherwise.
3. Always `diffBuilder.withPath()` and somehow extend it with some clever syntax to represent paths with objects as keys.
4. Extend the `diffBuilder` to be a lot more expressive in the type of errors it supports (e.g. it'd be cool if we could get errors like "Expected Set([1,2]) to contain 3.", or "Expected Map key Object({ x: [3] }) to be 42.", or some nice visual diffing with highlighting).

You probably have more ideas or things on your roadmap, so let me know what you think is best for this PR.